### PR TITLE
Add support for type-indexed Typeable

### PR DESCRIPTION
**Pull request by bgamari**

### CODE REVIEW

**Comment by phadej** [comment]
@bgamari there's a conflict

**Comment by bgamari** [comment]
The conflict has been slayed. I also took the liberty of removing a use of `bytestring`'s deprecated `inlinePerformIO` usage, which has been throwing warnings for quite some time now.

**Comment by bgamari** [comment]
Would it be okay to drop support for GHC 7.0 at this point?

**Comment by phadej** [comment]
I'd drop support for GHC 7.2 too as well, if we up to dropping support for older GHC.

**Comment by bgamari** [comment]
@tibbe, what do you say?

**Comment by tibbe** [comment]
We need to support the last 3 major releases of GHC.

File: Data/Hashable/Class.hs [hunk 1]
```diff
@@ -591,7 +601,7 @@ instance Hashable1 [] where
         step (SP s l) x   = SP (h s x) (l + 1)
 
 instance Hashable B.ByteString where
-    hashWithSalt salt bs = B.inlinePerformIO $
+    hashWithSalt salt bs = unsafeDupablePerformIO $
```
Comment (tibbe) [inline Data/Hashable/Class.hs]: Is this really performance equivalent? Could you run the benchmarks with and without this patch?

Comment (RyanGlScott) [inline Data/Hashable/Class.hs]: @tibbe, I checkout out the current HEAD state of `hashable` (what I will call "current" from here on out), as well as @bgamari's fork (what I will call "proposed"), and ran the benchmarks with both repos using a 4-core, 64-bit Linux desktop with 16 GB of RAM.

A preamble: it should be noted that the benchmarks currently do not compile. You have to make this change first before they will work:

```diff
diff --git a/hashable.cabal b/hashable.cabal
index 2dbf763..80cdf83 100644
--- a/hashable.cabal
+++ b/hashable.cabal
@@ -113,6 +113,7 @@ benchmark benchmarks
     base,
     bytestring,
     criterion >= 1.0,
+    deepseq >= 1.3,
     ghc-prim,
     siphash,
     text
```

Also, it should be noted that the benchmarks still contain uses of `inlinePerformIO` (not that that would affect the benchmark results, but it is worth noting if @bgamari's goal is to purge uses of `inlinePerformIO`).

Here are the results:

* [`bench-results-current.txt`](https://gist.githubusercontent.com/RyanGlScott/d4e30061a54cf05a23a8f43464f5070d/raw/893cad9bff101e1c9e4b837a11344c7a3a3b4da9/bench-results-current.txt)
* [`bench-results-proposed.txt`](https://gist.githubusercontent.com/RyanGlScott/d4e30061a54cf05a23a8f43464f5070d/raw/893cad9bff101e1c9e4b837a11344c7a3a3b4da9/bench-results-proposed.txt)

Happily, I can't spot any significant differences between the results for `current` and `proposed`.

**Comment by bgamari** [comment]
Can we merge this yet? 8.2 is imminent.

**Comment by ezyang** [comment]
Ran into this today!

**Comment by kolmodin** [comment]
I ran into this a couple of weeks ago, and then again today.

**Pull request merged**